src/index.js

```diff
@@ -31,31 +31,27 @@ class RpcEngine extends SafeEventEmitter {
   //
 
   async _handleBatch (reqs, cb) {
-    const batchRes = []
-    for (const r of reqs) {
-      try {
-        let [err, res] = await this._promiseHandle(r)
-        if (!res) {
-          if (err) {
-            throw err
-          } else {
-            throw ethErrors.rpc.internal('JsonRpcEngine: Request handler returned neither error nor response.')
-          }
-        } else {
-          batchRes.push(res)
-        }
-      } catch (_err) {
-        // some kind of fatal error
-        return cb(_err, null)
-      }
+
+    try {
+      const batchRes = await Promise.all(
+        reqs.map(this._promiseHandle.bind(this))
+      )
+      cb(null, batchRes)
+    } catch (err) {
+      cb(err)
     }
-    cb(null, batchRes)
   }
 
   _promiseHandle (req) {
-    return new Promise((resolve) => {
+    return new Promise((resolve, reject) => {
       this._handle(req, (err, res) => {
-        resolve([err, res])
+        if (!res) {
```

Comment:
Perhaps this should look for the existence of an error as well?
```suggestion
        if (err || !res) {
```

Reply:
No, if there is a response, it'll contain either a `result` or `error` property, and we want to add it to the `batchRes` array. OTOH, if there's no response at all, something bad has happened, and that's the case that condition is meant to catch.

Reply:
Hmm. That doesn't appear to be the case, from looking just at the `_handle` function.

```
  _handle (_req, cb) {
    // shallow clone request object
    const req = Object.assign({}, _req)
    // create response obj
    const res = {
      id: req.id,
      jsonrpc: req.jsonrpc
    }
    // process all middleware
    this._runMiddleware(req, res, (err) => {
      // take a clear any responseError
      const responseError = res._originalError
      delete res._originalError
      if (responseError) {
        // ensure no result is present on an errored response
        delete res.result
        // return originalError and response
        return cb(responseError, res)
      }
      // return response
      cb(err, res)
    })
  }
```

It checks for `_originalError`, and ensures the result isn't returned in that case, sure. But if `err` is truthy, it still returns the result as normal. That case is being missed here.

Reply:
I guess you're referring to somewhere else in the middleware stack; I'll take your word for it. It seems very unconventional though 😕

Reply:
Yeah, the full picture relies on behavior in `_runMiddleware`, which is my bad for not pointing out.

Reply:
Hmm, I think I see what you mean now. Well, can't say I like this pattern, but it certainly matches what's there before so it's correct as far as this PR is concerned 😄